### PR TITLE
Add declareOracleFun to the Java API

### PR DESCRIPTION
**Pull request by mudathirmahgoub**

### CODE REVIEW

File: src/api/java/io/github/cvc5/IOracle.java
Comment (abdoo8080) [inline src/api/java/io/github/cvc5/IOracle.java:20]: Format?

File: src/api/java/io/github/cvc5/IOracle.java
Comment (abdoo8080) [inline src/api/java/io/github/cvc5/IOracle.java:19]: Not a big fan of the word `compute`. Other suggestions:
- `run` (from `Runnable` interface)
- `apply` (from `Function<T,R> interface)

File: src/api/java/io/github/cvc5/IOracle.java
Comment (abdoo8080) [inline src/api/java/io/github/cvc5/IOracle.java:18]: If technically possible, I would like the APIs to be consistant. We should either use the same interface in all 3 APIs or use `std::function` equivalents in the Java and Python bindings. The Java equivalent to `std::function` is [`Function<T, R>`](https://docs.oracle.com/javase/8/docs/api/java/util/function/Function.html). I don't think Python has an equivalent, so an interface may be better.

File: src/api/java/io/github/cvc5/Solver.java
Comment (abdoo8080) [inline src/api/java/io/github/cvc5/Solver.java:2207]:
```suggestion
   * @param symbol The name of the oracle
```

**Review by abdoo8080 — COMMENTED** [review]
Thanks for volunteering to implement this. I don't know what would be the appropriate interface here, so I left some comments to start a discussion.

File: src/api/java/io/github/cvc5/IOracle.java
Comment (mudathirmahgoub) [reply src/api/java/io/github/cvc5/IOracle.java:18]: We can't use `Function<T, R>` here because the user is expected to call solver functions that may throw a CVC5ApiException.

File: src/api/java/io/github/cvc5/Solver.java
Comment (mudathirmahgoub) [reply src/api/java/io/github/cvc5/Solver.java:2207]: @ajreynol  This is copied from the cpp api documentation.

File: src/api/java/io/github/cvc5/IOracle.java
Comment (mudathirmahgoub) [reply src/api/java/io/github/cvc5/IOracle.java]: Renamed to `apply`

File: src/api/java/io/github/cvc5/IOracle.java
Comment (mudathirmahgoub) [reply src/api/java/io/github/cvc5/IOracle.java]: This is how clang-format formats java interface.

File: src/api/java/io/github/cvc5/Solver.java
Comment (abdoo8080) [reply src/api/java/io/github/cvc5/Solver.java:2207]: I remember adding this suggestion to Andy's PR too. I think it says `oracle` now.

File: src/api/java/io/github/cvc5/IOracle.java
Comment (4tXJ7f) [inline src/api/java/io/github/cvc5/IOracle.java:18]:
```suggestion
@FunctionalInterface
public interface IOracle {
```

File: src/api/java/io/github/cvc5/IOracle.java
Comment (4tXJ7f) [inline src/api/java/io/github/cvc5/IOracle.java:13]: Update comment

File: src/api/java/jni/solver.cpp
Comment (4tXJ7f) [inline src/api/java/jni/solver.cpp:2158]: To make future PRs easier to review, please don't change the formatting of unrelated code in the same PR.

File: src/api/java/jni/api_utilities.h
Comment (4tXJ7f) [inline src/api/java/jni/api_utilities.h:146]: We really need to rethink memory management, but let's leave this to future PRs.

**Review by 4tXJ7f — APPROVED** [review]
LGTM, just the comment in `IOracle.java` needs to be updated (marking as "do not merge" for this reason).

File: src/api/java/io/github/cvc5/IOracle.java
Comment (abdoo8080) [reply src/api/java/io/github/cvc5/IOracle.java:18]: Umm... How is that a problem? Isn't it the user's responsibility to deal with those exceptions? The internals do not expect an exception to be thrown when the oracle function is called...

File: src/api/java/io/github/cvc5/IOracle.java
Comment (mudathirmahgoub) [reply src/api/java/io/github/cvc5/IOracle.java:13]: Done.

File: src/api/java/jni/solver.cpp
Comment (mudathirmahgoub) [reply src/api/java/jni/solver.cpp:2158]: Unlike other cvc5 files, I made sure this file was clang-formatted from the start. 
Somehow, its format became broken with new commits. 
This means the flow used to format these files  is not working properly. 
The same applies for all java files, which I don't think are applied to new commits.

File: src/api/java/jni/api_utilities.h
Comment (mudathirmahgoub) [reply src/api/java/jni/api_utilities.h:146]: This was my main review point for this PR. I guess we can handle this later.

File: src/api/java/io/github/cvc5/IOracle.java
Comment (mudathirmahgoub) [reply src/api/java/io/github/cvc5/IOracle.java:21]: Yes, the user should handle the api exceptions. But I don't need to force the user to surround the body of each Oracle with try-catch block. 
For example, if the Functional<T, R> interface is used, the user needs to change replace this expression  
https://github.com/cvc5/cvc5/pull/8815/files#diff-0f2f1183d810770d1917a0457a53e85ddf5cb52c593c50798d8b63ac1ef6cc66R3020
with try catch block. 
Otherwise compile error would be thrown.

**Pull request merged**
